File: proga/1_semestr/labs/lab3/src/main/java/dev/lysmux/lab3/entity/people/ShortyGroup.java
package dev.lysmux.lab3.entity.people;

import dev.lysmux.lab3.common.Case;
import dev.lysmux.lab3.common.HasCase;

import java.util.Objects;

public class ShortyGroup extends Person implements HasCase {
    protected String state;

    public ShortyGroup(String state) {
        super("коротышки");
        this.state = state;
    }

    public ShortyGroup() {
        this("");
    }

    public String getState() {
        return state;
    }

    public void setState(String state) {
        this.state = state;
    }

    public String getDescription() {
        if (state.isEmpty()) return caseDeclension(Case.CREATIVE);
        return state + " " + caseDeclension(Case.CREATIVE);
    }

    @Override
    public String caseDeclension(Case to_case) {
        return switch (to_case) {
            case NOMINATIVE -> "коротышки";
            case GENITIVE, ACCUSATIVE -> "коротышек";
            case DATIVE -> "коротышкам";
            case CREATIVE -> "коротышками";
            case PREPOSITIONAL -> "коротышках";
        };
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ShortyGroup that)) return false;
        if (!super.equals(o)) return false;
        return Objects.equals(state, that.state);
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), state);
    }
}

File: proga/1_semestr/labs/lab3/src/main/kotlin/dev/lysmux/lab3/entity/people/ShortyGroup.kt
package dev.lysmux.lab3.entity.people

import dev.lysmux.lab3.common.Case
import dev.lysmux.lab3.common.HasCase

class ShortyGroup(private var state: String = "") : Person("коротышки"), HasCase{
    fun getState() = state
    fun setState(state: String) {
        this.state = state
    }

    fun getDescription(): String {
        if (state.isEmpty()) return caseDeclension(Case.CREATIVE)
        return "$state ${caseDeclension(Case.CREATIVE)}"
    }

    override fun caseDeclension(toCase: Case): String {
        return when (toCase) {
            Case.NOMINATIVE -> "коротышки"
            Case.ACCUSATIVE -> "коротышек"
            Case.GENITIVE -> "коротышек"
            Case.DATIVE -> "коротышкам"
            Case.CREATIVE -> "коротышками"
            Case.PREPOSITIONAL -> "коротышках"
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ShortyGroup) return false
        if (!super.equals(other)) return false

        return state == other.state
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + state.hashCode()
        return result
    }
}